urllib3/contrib/appenginemanager.py

```diff
@@ -0,0 +1,117 @@
+import warnings
+
+from ..exceptions import (
+    HTTPError,
+    HTTPWarning,
+    ProtocolError,
+    TimeoutError,
+    SSLError
+)
+
+from ..packages.six import BytesIO
+from ..request import RequestMethods
+from ..response import HTTPResponse
+from ..util.timeout import Timeout
+
+try:
+    from google.appengine.api import urlfetch
+except ImportError:
+    urlfetch = None
+
+
+class AppEnginePlatformWarning(HTTPWarning):
+    pass
+
+
+class AppEnginePlatformError(HTTPError):
+    pass
+
+
+class AppEngineManager(RequestMethods):
+    """
+    Connection manager for Google App Engine sandbox applications.
+
+    This manager uses the URLFetch service directly instead of using the
+    emulated httplib.
+    """
+
+    def __init__(self, headers=None, validate_certificate=True):
+
+        if not urlfetch:
+            raise AppEnginePlatformError(
+                "URLFetch is not available in this environment.")
+
+        warnings.warn(
+            "urllib3 is using URLFetch on Google App Engine sandbox instead "
+            "of sockets. To use sockets directly instead of URLFetch see "
+            "<TODO: documentation link>.", AppEnginePlatformWarning)
+
+        RequestMethods.__init__(self, headers)
+        self.validate_certificate = validate_certificate
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # Return False to re-raise any potential exceptions
+        return False
+
+    def urlopen(self, method, url, body=None, headers=None,
+                redirect=True, timeout=Timeout.DEFAULT_TIMEOUT,
+                **response_kw):
+
+        try:
+            response = urlfetch.fetch(
+                url,
+                payload=body,
+                method=method,
+                headers=headers or {},
+                allow_truncated=False,  # revisit
+                follow_redirects=redirect,
+                deadline=self._get_absolute_timeout(timeout),
+                validate_certificate=self.validate_certificate,
+            )
+        except urlfetch.DeadlineExceededError as e:
+            raise TimeoutError(self, e)
+
+        except urlfetch.InvalidURLError as e:
+            if 'too large' in e.message:
+                warnings.warn("URLFetch request too large, URLFetch only "
+                              "supports requests up to 10mb in size.",
+                              AppEnginePlatformWarning)
+            raise ProtocolError(e)
+
+        except urlfetch.DownloadError as e:
+            raise ProtocolError(e)
+
+        except urlfetch.ResponseTooLargeError as e:
+            warnings.warn("URLFetch response too large, URLFetch only supports"
```

Comment:
Up to you. So far I've tried to maintain compatibility with URLFetch's API. If you feel it would be more appropriate to raise the AppEnginePlatformError, that's fine with me.